bigtable/client/table_readrow_test.cc

```diff
@@ -0,0 +1,95 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "bigtable/client/table.h"
+#include "bigtable/client/testing/table_test_fixture.h"
+
+namespace {
+class MockResponseStream : public grpc::ClientReaderInterface<
+                               ::google::bigtable::v2::ReadRowsResponse> {
+ public:
+  MOCK_METHOD0(WaitForInitialMetadata, void());
+  MOCK_METHOD0(Finish, grpc::Status());
+  MOCK_METHOD1(NextMessageSize, bool(std::uint32_t *));
+  MOCK_METHOD1(Read, bool(::google::bigtable::v2::ReadRowsResponse *));
+};
+
+/// Define helper types and functions for this test.
+class TableReadRowTest : public bigtable::testing::TableTestFixture {};
+
+}  // anonymous namespace
+
+TEST_F(TableReadRowTest, ReadRowSimple) {
+  using namespace ::testing;
+  namespace btproto = ::google::bigtable::v2;
+
+  auto response = bigtable::testing::ReadRowsResponseFromString(R"(
+      chunks {
+        row_key: "r1"
+        family_name { value: "fam" }
+        qualifier { value: "col" }
+        timestamp_micros: 42000
+        value: "value"
+        commit_row: true
+      }
+)");
+
+  // must be a new pointer, it is wrapped in unique_ptr by ReadRows
+  auto stream = absl::make_unique<MockResponseStream>();
+  EXPECT_CALL(*stream, Read(_))
+      .WillOnce(Invoke([&response](btproto::ReadRowsResponse *r) {
+        *r = response;
+        return true;
+      }))
+      .WillOnce(Return(false));
+  EXPECT_CALL(*stream, Finish()).WillOnce(Return(grpc::Status::OK));
+
+  EXPECT_CALL(*bigtable_stub_, ReadRowsRaw(_, _))
+      .WillOnce(Invoke([&stream, this](grpc::ClientContext *,
+                                       btproto::ReadRowsRequest const &req) {
+        EXPECT_EQ(1, req.rows().row_keys_size());
```

Comment:
Nice, I didn't realize the EXPECT_* macros work in lambdas as well. I wrote some silly matchers in my tests.

Reply:
`EXPECT_EQ()` works fine inside lambdas.  `ASSERT_EQ()` does not always work because it expands to `if (blah) { return; }`, and the lambda may return something other than `void`.  `googletest` should use exceptions :grin: